File: src/config/paths.ts
import { homedir } from 'node:os';
import { join } from 'node:path';

const appDir = join(homedir(), '.feishu-codex-bridge');
const larkCliDir = join(appDir, 'lark-cli');
const codexCliDir = join(appDir, 'codex-cli');

/**
 * Per-bot state directory. Each saved bot keeps its own config / projects /
 * sessions / single-instance lock under `~/.feishu-codex-bridge/bots/<appId>/`
 * so switching the active bot (`use`) never mixes one bot's groups with
 * another's. `currentBotDir` defaults to `appDir` (the legacy flat layout) so
 * code that runs before a bot is selected — and pre-migration installs — keeps
 * reading the old top-level files; `useBotDir()` repoints it once the active
 * bot is known.
 */
let currentBotDir = appDir;

export function botDir(appId: string): string {
  return join(appDir, 'bots', appId);
}

/** Point the per-bot paths at `appId`'s directory. Call once at startup.
 * ⚠️ daemon 进程内（run/supervisor）绝不可在请求路径上反复调它切目录——它是
 * 模块级全局态，会把在跑 bot 进程的 paths 指到别的 bot。跨 bot 聚合读取一律
 * 走 {@link botPaths} 的显式路径。 */
export function useBotDir(appId: string): void {
  currentBotDir = botDir(appId);
}

/** 指定 bot 的各状态文件路径（纯函数，不碰全局 currentBotDir）。Web 控制台 /
 * supervisor 聚合多 bot 读取专用——与 useBotDir 后的 paths.* 指向完全一致。 */
export function botPaths(appId: string): {
  dir: string;
  configFile: string;
  sessionsFile: string;
  projectsFile: string;
  processesFile: string;
} {
  const dir = botDir(appId);
  return {
    dir,
    configFile: join(dir, 'config.json'),
    sessionsFile: join(dir, 'sessions.json'),
    projectsFile: join(dir, 'projects.json'),
    processesFile: join(dir, 'processes.json'),
  };
}

export const paths = {
  appDir,
  cacheDir: appDir,
  /** bot 注册表：保存的全部 bot + 当前选中的 appId */
  botsFile: join(appDir, 'bots.json'),
  /** app id / 租户 / 偏好（当前 bot；不含明文密钥） */
  get configFile(): string {
    return join(currentBotDir, 'config.json');
  },
  /** thread(话题) → codex thread_id + cwd + 会话级配置（当前 bot） */
  get sessionsFile(): string {
    return join(currentBotDir, 'sessions.json');
  },
  /** project(群) → cwd + 默认参数 注册表（当前 bot） */
  get projectsFile(): string {
    return join(currentBotDir, 'projects.json');
  },
  /** 在跑的 start 进程注册中心（同 App 冲突检测；当前 bot） */
  get processesFile(): string {
    return join(currentBotDir, 'processes.json');
  },
  secretsFile: join(appDir, 'secrets.enc'),
  keystoreSaltFile: join(appDir, '.keystore.salt'),
  npmCacheDir: join(appDir, 'npm-cache'),
  /**
   * 按需重后端（claude-sdk 等 npm-ondemand 包）私装目录：一个扁平
   * `~/.feishu-codex-bridge/backends/node_modules` 放所有按需后端的 npm 包。
   * 永远在用户 HOME 下、用户可写（零 sudo/brew），与全局包目录的权限死结解耦。
   * 解析靠 createRequire(backendsDir/...).resolve（见 agent/backend-loader）；
   * 安装靠 `npm install --prefix backendsDir`（见 agent/installer）。 */
  backendsDir: join(appDir, 'backends'),
  /** 空白项目默认落地目录 */
  projectsRootDir: join(appDir, 'projects'),
  larkCliDir,
  larkCliBinDir: join(larkCliDir, 'node_modules', '.bin'),
  codexCliDir,
  codexCliBinDir: join(codexCliDir, 'node_modules', '.bin'),
  /**
   * Thin shell wrapper that lark-cli invokes to resolve secrets from the
   * bridge's encrypted store. Written user-owned and non-symlinked so it
   * passes lark-cli's AssertSecurePath audit.
   */
  secretsGetterScript: join(appDir, 'secrets-getter'),
  mediaDir: join(appDir, 'media'),
  /** Inbound file attachments downloaded from chat, handed to codex by absolute
   * path (codex has no native file input). TTL-pruned like {@link mediaDir}. */
  inboundDir: join(appDir, 'inbound'),
  /** daemon 内嵌 Web 控制台的发现文件 {port, token, pid}（0600，daemon 退出
   * 清理）——`web` 子命令据此直接打开 daemon 控制台而不是再起只读副本。 */
  webConsoleFile: join(appDir, 'web-console.json'),
};

File: internal/config/paths.go
package config

import (
	"os"
	"os/user"
	"path/filepath"
	"strings"
)

// paths.go —— 全部文件系统路径常量（对齐 TS config/paths，见方案 §4）。
//
// Go 改进（相对 TS）：
//   - TS 用模块级 currentBotDir 全局态切目录（注释警告 daemon 不可在请求路径上调）。
//     Go 版改为 Paths{AppID} 实例 + 纯函数 BotXxxFile(appID)，显式传 appId，消除全局态。
//   - 全局文件（secrets.enc / bots.json / logs/ 等）用包级函数；per-bot 文件用 Paths 或纯函数。

const appDirName = ".feishu-codex-bridge"

// AppDir 返回 ~/.feishu-codex-bridge。
func AppDir() string {
	home, err := os.UserHomeDir()
	if err != nil || home == "" {
		home = os.Getenv("HOME")
	}
	return filepath.Join(home, appDirName)
}

// ── 全局文件（与具体 bot 无关）──────────────────────────────────

func BotsFile() string            { return filepath.Join(AppDir(), "bots.json") }
func SecretsFile() string         { return filepath.Join(AppDir(), "secrets.enc") }
func KeystoreSaltFile() string    { return filepath.Join(AppDir(), ".keystore.salt") }
func SecretsGetterScript() string { return filepath.Join(AppDir(), "secrets-getter") }
func LogsDir() string             { return filepath.Join(AppDir(), "logs") }
func ServiceLog() string          { return filepath.Join(AppDir(), "service.log") }
func ServiceErrLog() string       { return filepath.Join(AppDir(), "service.err.log") }
func MediaDir() string            { return filepath.Join(AppDir(), "media") }
func InboundDir() string          { return filepath.Join(AppDir(), "inbound") }
func BackendsDir() string         { return filepath.Join(AppDir(), "backends") }
func ProjectsRootDir() string     { return filepath.Join(AppDir(), "projects") }

// ResolveProjectsRootDir 取空白项目默认父目录：偏好 ProjectsRootDir 优先，否则 AppDir()/projects。
// 支持 ~ 开头与绝对路径；相对路径按 AppDir 解析。
func ResolveProjectsRootDir(cfg AppConfig) string {
	root := ""
	if cfg.Preferences != nil {
		root = strings.TrimSpace(cfg.Preferences.ProjectsRootDir)
	}
	if root == "" {
		return ProjectsRootDir()
	}
	if strings.HasPrefix(root, "~") {
		if home, err := os.UserHomeDir(); err == nil && home != "" {
			root = filepath.Join(home, strings.TrimPrefix(root, "~"))
		}
	}
	if !filepath.IsAbs(root) {
		root = filepath.Join(AppDir(), root)
	}
	return root
}
func WebConsoleFile() string      { return filepath.Join(AppDir(), "web-console.json") }
func WebTokenFile() string        { return filepath.Join(AppDir(), "web-token") }
func DaemonPIDFile() string       { return filepath.Join(AppDir(), "daemon.pid") }

// ── per-bot 路径（纯函数，不碰全局态）───────────────────────────

func BotDir(appID string) string           { return filepath.Join(AppDir(), "bots", appID) }
func BotConfigFile(appID string) string    { return filepath.Join(BotDir(appID), "config.json") }
func BotSessionsFile(appID string) string  { return filepath.Join(BotDir(appID), "sessions.json") }
func BotProjectsFile(appID string) string  { return filepath.Join(BotDir(appID), "projects.json") }
func BotProcessesFile(appID string) string { return filepath.Join(BotDir(appID), "processes.json") }

// BotCommentInstructionsFile 云文档评论 @bot 的可编辑提示词 master 文件（当前 bot）。
// 用户直接编辑这一份，桥在每条评论运行前把它同步进该文档的评论工作目录（AGENTS.md / CLAUDE.md）。
// 首次缺失时由 comments 模块用内置默认模板自动落地（对齐 TS commentInstructionsFile）。
func BotCommentInstructionsFile(appID string) string {
	return filepath.Join(BotDir(appID), "comment-instructions.md")
}

// BotCommentsRootDir 评论工作目录根（当前 bot）：每个被评论文档一个 comment-<type>-<token> 子目录，
// 放同步进去的 AGENTS.md / CLAUDE.md。per-bot 隔离，避免编辑提示词时越界改到别的 bot（对齐 TS commentsRootDir）。
func BotCommentsRootDir(appID string) string {
	return filepath.Join(BotDir(appID), "comments")
}

// Paths 持有某 bot 的路径集合，替代 TS 的全局 currentBotDir。
// 启动时按活跃 bot 构造一份；跨 bot 聚合用上面的纯函数。
type Paths struct {
	AppID string
}

func NewPaths(appID string) *Paths { return &Paths{AppID: appID} }

func (p *Paths) Dir() string           { return BotDir(p.AppID) }
func (p *Paths) ConfigFile() string    { return BotConfigFile(p.AppID) }
func (p *Paths) SessionsFile() string  { return BotSessionsFile(p.AppID) }
func (p *Paths) ProjectsFile() string  { return BotProjectsFile(p.AppID) }
func (p *Paths) ProcessesFile() string { return BotProcessesFile(p.AppID) }
func (p *Paths) CommentInstructionsFile() string { return BotCommentInstructionsFile(p.AppID) }
func (p *Paths) CommentsRootDir() string         { return BotCommentsRootDir(p.AppID) }

// CliBridgeSocket 返回 cli-bridge 的 Unix socket 路径（二期 cli-bridge 用）。
func (p *Paths) CliBridgeSocket() string { return filepath.Join(p.Dir(), "cli-bridge.sock") }

// KeystoreSeed 返回 keystore 派生种子 "hostname|username"（与 TS 字节级一致）。
// 测试用 Keystore.WithSeed 注入固定值，无需依赖真实主机名。
func KeystoreSeed() string {
	host, _ := os.Hostname()
	username := ""
	if u, err := user.Current(); err == nil && u != nil {
		username = u.Username
	}
	return host + "|" + username
}
